src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -227,24 +228,99 @@ public int capacity() {
      * For positive values of {@code minimumCapacity}, this method
-     * behaves like {@code ensureCapacity}, however it is never
-     * synchronized.
-     * If {@code minimumCapacity} is non positive due to numeric
+     * behaves like {@code ensureCapacity}, however it is never synchronized.
+     * If {@code minimumCapacity} is non-positive due to numeric
      * overflow, this method throws {@code OutOfMemoryError}.
+     * @param value the current buffer
+     * @param coder of the current buffer
+     * @param count the count of chars in the current buffer
+     * @param minimumCapacity the new minimum capacity
+     * @param newCoder the desired new coder
+     */
+    private static byte[] ensureCapacityNewCoder(byte[] value, byte coder, int count,
+                                                 int minimumCapacity, byte newCoder) {
+        assert coder == newCoder || newCoder == UTF16 : "bad new coder UTF16 -> LATIN1";
+        // overflow-conscious code
+        int growth = minimumCapacity - (value.length >> coder);
+        if (coder == newCoder) {
+            if (growth > 0) {
+                // copy all bytes to new larger buffer
+                value = Arrays.copyOf(value,
+                        newCapacity(value, newCoder, minimumCapacity) << newCoder);
+            }
+            return value;
+        } else {
+            // inflate (and grow if additional length is requested)
+            // always growing might be better but it breaks current Capacity tests
+            int newCap = (growth > 0)
+                ? newCapacity(value, newCoder, minimumCapacity)
+                : value.length;
+            byte[] newValue = StringUTF16.newBytesFor(newCap);
+            StringLatin1.inflate(value, 0, newValue, 0, count);
+            return newValue;
+        }
+    }
+
+    /**
+     * {@return the value buffer sufficient to hold the minimumCapactity and a copy of the contents}
+     * There is no change to the coder.
+     * The current value buffer is returned if the size is already sufficient.
+     * For positive values of {@code minimumCapacity}, this method
+     * behaves like {@code ensureCapacity}, however it is never synchronized.
+     * If {@code minimumCapacity} is non-positive due to numeric
+     * overflow, this method throws {@code OutOfMemoryError}.
+     * @param value the current buffer
+     * @param coder of the current buffer
+     * @param minimumCapacity the new minimum capacity
      */
-    private void ensureCapacityInternal(int minimumCapacity) {
+    private static byte[] ensureCapacitySameCoder(byte[] value, byte coder, int minimumCapacity) {
         // overflow-conscious code
         int oldCapacity = value.length >> coder;
         if (minimumCapacity - oldCapacity > 0) {
             value = Arrays.copyOf(value,
-                    newCapacity(minimumCapacity) << coder);
+                    newCapacity(value, coder, minimumCapacity) << coder);
```

Comment:
Same question about whether we should assert that the new length we determine for the buffer is not lesser than the current length.